src/stm32f1/adc.cpp

```diff
@@ -0,0 +1,177 @@
+// Copyright 2024 Khalil Estell
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <libhal-stm32f1/adc.hpp>
+#include <libhal-stm32f1/clock.hpp>
+#include <libhal-stm32f1/constants.hpp>
+#include <libhal-util/bit.hpp>
+#include <libhal-util/bit_limits.hpp>
+
+#include "adc_reg.hpp"
+#include "pin.hpp"
+#include "power.hpp"
+
+namespace hal::stm32f1 {
+
+namespace {
+void setup(adc::channel const& p_channel)
+{
+  auto adc_frequency = frequency(hal::stm32f1::peripheral::adc1);
```

Comment:
```suggestion
  auto const adc_frequency = frequency(hal::stm32f1::peripheral::adc1);
```

Add const to variables that you know will not change.